grakn-graph/src/main/java/ai/grakn/graph/internal/TypeImpl.java

```diff
@@ -321,17 +324,31 @@ protected void checkTypeMutation(){
      * @return The resulting relation type which allows instances of this type to have relations with the provided resourceType.
      */
     public RelationType hasResource(ResourceType resourceType, boolean required){
-        String resourceTypeId = resourceType.getName();
-        RoleType ownerRole = getGraknGraph().putRoleTypeImplicit(Schema.Resource.HAS_RESOURCE_OWNER.getName(resourceTypeId));
-        RoleType valueRole = getGraknGraph().putRoleTypeImplicit(Schema.Resource.HAS_RESOURCE_VALUE.getName(resourceTypeId));
+        String resourceTypeName = resourceType.getName();
+        RoleType ownerRole = getGraknGraph().putRoleTypeImplicit(Schema.Resource.HAS_RESOURCE_OWNER.getName(resourceTypeName));
+        RoleType valueRole = getGraknGraph().putRoleTypeImplicit(Schema.Resource.HAS_RESOURCE_VALUE.getName(resourceTypeName));
+        RelationType relationType = getGraknGraph().putRelationTypeImplicit(Schema.Resource.HAS_RESOURCE.getName(resourceTypeName)).
+                hasRole(ownerRole).
+                hasRole(valueRole);
+
+        //Linking with ako structure if present
+        ResourceType resourceTypeSuper = resourceType.superType();
+        if(resourceTypeSuper != null){
+            String superName = resourceTypeSuper.getName();
+            RoleType ownerRoleSuper = getGraknGraph().putRoleTypeImplicit(Schema.Resource.HAS_RESOURCE_OWNER.getName(superName));
+            RoleType valueRoleSuper = getGraknGraph().putRoleTypeImplicit(Schema.Resource.HAS_RESOURCE_VALUE.getName(superName));
+            RelationType relationTypeSuper = getGraknGraph().putRelationTypeImplicit(Schema.Resource.HAS_RESOURCE.getName(superName)).
+                    hasRole(ownerRoleSuper).hasRole(valueRoleSuper);
+
+            ownerRole.superType(ownerRoleSuper);
+            valueRole.superType(valueRoleSuper);
+            relationType.superType(relationTypeSuper);
+        }
```

Comment:
Horrible edge case: what if the supertype is _also_ a subtype of another resource-type? It will not get its `superType` set on its new resource-type and role-types here.